baseapp/baseapp.go

```diff
@@ -142,7 +142,14 @@ type BaseApp struct { //nolint: maligned
 
 	// abciListeners for hooking into the ABCI message processing of the BaseApp
 	// and exposing the requests and responses to external consumers
-	abciListeners []storetypes.ABCIListener
+	abciListeners []ABCIListener
+
+	// abciListenersAsync for determining if abciListeners will run asynchronously.
+	abciListenersAsync bool
+
+	// stopNodeOnABCIListenerErr halts the node when ABCI streaming service listening results in an error.
+	// stopNodeOnABCIListenerErr=true MUST be paired with abciListenersAsync=false, otherwise it will be ignored.
+	stopNodeOnABCIListenerErr bool
```

Comment:
Why not have this all be part of a single `StreamingManager` type? This manager could be responsible for all writes and cleanup.

Reply:
I agree, I'm looking at the best way to push it down to the `ListenBeginBlock`. If you take a look the `NewStreamingPlugin ` function, it serves plugins by doing a lookup through the `PluginMap`, which is a struct, and the function itself is `NewStreamingPlugin`. One way I'm looking at is to pass it down through the `ctx`. This would require updating `setState` (see below) and perhaps adding a `StreamingManager` function to the `sdk.Context{}`?


https://github.com/cosmos/cosmos-sdk/blob/3ed0fedb509749c14f96f71cf4ba3ce031539314/streaming/streaming.go#L23-L31

What I'm currently testing with the below. It needs to be set for every new context.
```go
func (app *BaseApp) setState(mode runTxMode, header tmproto.Header) {
	...
	baseState.ctx = baseState.ctx.WithValue("stopNodeOnABCIListenerErr", app.stopNodeOnABCIListenerErr)
        ...
```

perhaps the ^^^ could become?

```go
func (app *BaseApp) setState(mode runTxMode, header tmproto.Header) {
	...
        baseState.ctx = baseState.ctx.WithStreamingManager(app.streamingManager)
        ...
```

Reply:
this relates to the question from @kocubinski as well https://github.com/cosmos/cosmos-sdk/pull/14207#discussion_r1066025287.

Reply:
@alexanderbez @kocubinski take a look at it now and let me know what you think. I've grouped them into a `StreamingManager` and moved the cleanup and exit to the plugin layer.

Reply:
Personally I find this approach has a better separation of concerns :clap: